Add myaddr dynamic dns provider

brianshea2 opened the pull request:
Adds new dynamic dns provider: https://myaddr.tools
Partially fixes #728 (additional PR for public IP provider to come)
Cheers!

qdm12 commented on cmd/ddns-updater/main.go:
```diff
@@ -167,6 +167,14 @@ func _main(ctx context.Context, reader *reader.Reader, args []string, logger log
 	client := &http.Client{Timeout: config.Client.Timeout}
 	defer client.CloseIdleConnections()
 
+	for _, p := range providers {
+		err = provider.Init(p, ctx, client)
+		if err != nil {
+			shoutrrrClient.Notify(err.Error())
+			return err
+		}
+	}
+
```
Let's not do this, just force the user to set a domain name in the config.json, to match the existing architecture. There are also issues where for example the addrtools API does not work, then we won't know what domain name is failing (just the key) which isn't great.

qdm12 commented on internal/params/json.go:
```diff
@@ -171,11 +171,13 @@ func makeSettingsFromObject(common commonSettings, rawSettings json.RawMessage,
 			fmt.Sprintf("you can specify the owner %q directly in the domain field as %q",
 				common.Owner, strings.Join(domains, ",")))
-	} else { // extract owner(s) from domain(s)
+	} else if common.Domain != "" { // extract owner(s) from domain(s)
 		domain, owners, err = extractFromDomainField(common.Domain)
 		if err != nil {
 			return nil, nil, fmt.Errorf("extracting owners from domains: %w", err)
 		}
+	} else { // domain(s) not specified
+		owners = make([]string, 1) // required to enter provider initialization loop below
```
Related to init: revert this once the init function is removed 😉

qdm12 commented on internal/provider/provider.go:
```diff
@@ -75,6 +76,10 @@ type Provider interface {
 	Update(ctx context.Context, client *http.Client, ip netip.Addr) (newIP netip.Addr, err error)
 }
 
+type Initializer interface {
+	Init(ctx context.Context, client *http.Client) error
+}
+
```
Related to init: remove this

qdm12 commented on internal/provider/provider.go:
```diff
@@ -186,3 +193,10 @@ func New(providerName models.Provider, data json.RawMessage, domain, owner strin
 		return nil, fmt.Errorf("%w: %s", ErrProviderUnknown, providerName)
 	}
 }
+
+func Init(provider Provider, ctx context.Context, client *http.Client) error {
+	if initializer, ok := provider.(Initializer); ok {
+		return initializer.Init(ctx, client)
+	}
+	return nil
+}
```
Related to init: remove

qdm12 commented on internal/provider/providers/myaddr/provider.go:
```diff
@@ -0,0 +1,178 @@
+		Key string `json:"key"`
+	}
+	if err = json.Unmarshal(data, &providerSpecificSettings); err != nil {
+		err = fmt.Errorf("json decoding provider specific settings (myaddr): %w", err)
+		return
```
```suggestion
		return nil, fmt.Errorf("json decoding provider specific settings: %w", err)
```

Also no point having the provider name, that's wrapped by calling layers (or should be, if not!)

qdm12 commented on internal/provider/providers/myaddr/provider.go:
```diff
@@ -0,0 +1,178 @@
+		return
+	}
+	if providerSpecificSettings.Key == "" {
+		err = fmt.Errorf("validating provider specific settings (myaddr): %w", errors.ErrKeyNotSet)
+		return
```
```suggestion
		return nil, fmt.Errorf("validating provider specific settings: %w", errors.ErrKeyNotSet)
```

qdm12 commented on internal/provider/providers/myaddr/provider.go:
```diff
@@ -0,0 +1,178 @@
+		err = fmt.Errorf("validating provider specific settings (myaddr): %w", errors.ErrKeyNotSet)
+		return
+	}
+	p = &Provider{
+		key:        providerSpecificSettings.Key,
+		ipVersion:  ipVersion,
+		ipv6Suffix: ipv6Suffix,
+	}
+	return
```
```suggestion
	return &Provider{
		key:        providerSpecificSettings.Key,
		ipVersion:  ipVersion,
		ipv6Suffix: ipv6Suffix,
	}, nil
```

qdm12 commented on internal/provider/providers/myaddr/provider.go:
```diff
@@ -0,0 +1,178 @@
+	return
+}
+
+func (p *Provider) Init(ctx context.Context, client *http.Client) (err error) {
```
remove entire function 🙏 Thanks for trying to optimize this though!

qdm12 commented on internal/provider/providers/myaddr/provider.go:
```diff
@@ -0,0 +1,178 @@
+	v.Set("key", p.key)
+	v.Set("ip", ip.String())
+	b := strings.NewReader(v.Encode())
+	var req *http.Request
+	if req, err = http.NewRequestWithContext(ctx, http.MethodPost, u.String(), b); err != nil {
+		err = fmt.Errorf("creating http request: %w", err)
+		return
+	}
```
no short ifs (more readable), no naked returns (otherwise it's hard to know what's being returned), and let's use full name variable names `req` -> `request` for better readability

```suggestion
	request, err := http.NewRequestWithContext(ctx, http.MethodPost, u.String(), b)
	if err != nil {
		return netip.Addr{}, fmt.Errorf("creating http request: %w", err)
	}
```

qdm12 commented on internal/provider/providers/myaddr/provider.go:
```diff
@@ -0,0 +1,178 @@
+	headers.SetContentType(req, "application/x-www-form-urlencoded")
+	headers.SetUserAgent(req)
+
+	// execute HTTP request
```
unneeded comment, the code speaks for itself below

```suggestion
```

qdm12 commented on internal/provider/providers/myaddr/provider.go:
```diff
@@ -0,0 +1,178 @@
+	headers.SetUserAgent(req)
+
+	// execute HTTP request
+	var resp *http.Response
+	if resp, err = client.Do(req); err != nil {
+		return
+	}
```
```suggestion
	response, err := client.Do(request)
	if err != nil {
		return netip.Addr{}, err
	}
```

qdm12 commented on internal/provider/providers/myaddr/provider.go:
```diff
@@ -0,0 +1,178 @@
+		return
+	}
+	defer resp.Body.Close()
+	defer io.Copy(io.Discard, resp.Body) // ensure body is read to completion
```
Interesting, but unneeded, see https://gist.github.com/mholt/eba0f2cc96658be0f717?permalink_comment_id=2605879#gistcomment-2605879

```suggestion
```

qdm12 commented on internal/provider/providers/myaddr/provider.go:
```diff
@@ -0,0 +1,178 @@
+	defer resp.Body.Close()
+	defer io.Copy(io.Discard, resp.Body) // ensure body is read to completion
+
+	// handle HTTP response
+	switch resp.StatusCode {
+	case http.StatusOK:
+		// update was successful
+		newIP = ip
+	case http.StatusBadRequest:
+		// bad request
+		err = fmt.Errorf("%w: %s", errors.ErrBadRequest, utils.BodyToSingleLine(resp.Body))
+	case http.StatusNotFound:
+		// registration not found
+		err = fmt.Errorf("%w: %s", errors.ErrKeyNotValid, utils.BodyToSingleLine(resp.Body))
+	default:
+		// unexpected HTTP status
+		err = fmt.Errorf("%w: %d: %s",
+			errors.ErrHTTPStatusNotValid, resp.StatusCode, utils.BodyToSingleLine(resp.Body))
+	}
+	return
```
- some unneeded comments
- no naked returns
- return early, almost always, easier to read

```suggestion
	switch response.StatusCode {
	case http.StatusOK:
		return ip, nil
	case http.StatusBadRequest:
		return netip.Addr{}, fmt.Errorf("%w: %s", errors.ErrBadRequest, utils.BodyToSingleLine(response.Body))
	case http.StatusNotFound
		return netip.Addr{}, fmt.Errorf("%w: %s", errors.ErrKeyNotValid, utils.BodyToSingleLine(response.Body))
	default:
		return netip.Addr{}, fmt.Errorf("%w: %d: %s",
			errors.ErrHTTPStatusNotValid, response.StatusCode, utils.BodyToSingleLine(response.Body))
	}
```

qdm12 commented on internal/provider/providers/myaddr/provider.go:
```diff
@@ -0,0 +1,178 @@
+	v := url.Values{}
+	v.Set("key", p.key)
+	v.Set("ip", ip.String())
+	b := strings.NewReader(v.Encode())
```
nit `b` -> `buffer` would be nice 😉

qdm12 commented on internal/provider/providers/myaddr/provider.go:
```diff
@@ -0,0 +1,178 @@
+}
+
+func (p *Provider) Update(ctx context.Context, client *http.Client, ip netip.Addr) (newIP netip.Addr, err error) {
+	// generate HTTP request
```
uneeded comment

```suggestion
```

qdm12 reviewed (changes requested):
Thanks for the PR 🎖️ 💯 !

A few comments of things to change here and there, but before jumping in this: would you mind creating another separate pull request for the public ip feature, since it should be a separate commit/PR really. Thank you 🙏 👍

brianshea2 commented:
Thank you for the quick review :) I believe I've addressed all your comments and made the code more consistent with the project.

I've reverted the public IP provider addition for now and will open a separate PR for that.

I understand your want for providing the `"domain"` to be consistent with other providers, and have removed the initialization code which fetched the domain corresponding to the provided `"key"`. My concern now is there's nothing explicitly linking the provided `"key"` and `"domain"`. I believe I've described the `"domain"` adequately in docs/myaddr.md to this point. But to reiterate, the provided `"key"` is the only identifier sent to the provider api, and is used to update all `[*.]name-corresponding-to-key.myaddr.{tools,dev,io}` in one fell swoop, which may or may not agree with the separately-user-provided `"domain"`, now used only for checking if an update is required. What are your thoughts here? It's possible to perform an additional api request during each update to fetch the domain corresponding to `"key"` and compare to the provided `"domain"` if wanted.

qdm12 commented:
> I believe I've addressed all your comments and made the code more consistent with the project.

Awesome 💯 

> I've reverted the public IP provider addition for now and will open a separate PR for that.

Perfect 🥇 

> It's possible to perform an additional api request during each update to fetch the domain corresponding to "key" and compare to the provided "domain" if wanted.

Eh it's a bit on the user if he/she sets a key not matching the domain. We can do foolproof checks as much as possible, but this seems like a bit too much at the cost of over-complicating it 😄 The API will probably just say the key is wrong anyway.

I did however add a little check to prevent multiple domains to be defined in a myaddr settings block since one key is for one domain (and its subdomains), so there would be no point handling it as different entries/updates in parallel, except getting banned for sending too many update requests 😄 

Merging this !!
